hw/ip/otbn/util/make_mulqacc.py

```diff
@@ -0,0 +1,272 @@
+
+    # Ensure that the first instruction zeroes the accumulator.
+    lines[0].modifiers.append('z')
+
+    return lines
+
+
+def format_block(lines: List[Line]) -> List[str]:
+    '''Format the lines so that the left-hand operands are all aligned.'''
+    align = max([line.min_lhs_align() for line in lines])
+    return [line.pretty(align) for line in lines]
+
+
+def is_wdr(reg_name: str) -> bool:
+    if not reg_name.startswith('w'):
+        return False
+    if len(reg_name) < 1:
+        return False
+    reg_num = reg_name[1:]
+    if not reg_num.isnumeric():
+        return False
+    reg_idx = int(reg_num)
+    return (0 <= reg_idx) and (reg_idx <= 31)
+
+
+def check_regs(regs: List[str], bits: int) -> None:
+    '''Checks that register names are valid and there are enough registers to
+       store the given number of bits.'''
+    for reg_name in regs:
+        if not is_wdr(reg_name):
+            raise ValueError(f'{reg_name} is not a valid wide data register.')
+    if len(regs) * WDR_SIZE < bits:
+        raise ValueError(
+            f'Register list {regs} does not have enough space to store a '
+            f'{bits}-bit number.')
+    return
+
+
+def main() -> int:
+    parser = argparse.ArgumentParser(description=(
+        'Print out a bn.mulqacc block for the given bignum product.'))
+    parser.add_argument('--lhs-bits',
+                        type=int,
+                        required=True,
+                        help=('Number of bits in the left-hand operand.'))
+    parser.add_argument(
+        '--lhs-regs',
+        nargs='+',
+        type=str,
+        required=True,
+        help=('Registers for left-hand operand (little-endian order).'))
+    parser.add_argument('--rhs-bits',
+                        type=int,
+                        required=True,
+                        help=('Number of bits in the right-hand operand.'))
+    parser.add_argument(
+        '--rhs-regs',
+        nargs='+',
+        type=str,
+        required=True,
+        help=('Registers for the right-hand operand (little-endian order).'))
+    parser.add_argument(
+        '--result-regs',
+        nargs='+',
+        type=str,
+        required=True,
+        help=('Registers in which to store the result. Must not overlap with '
+              'operand registers.'))
+    args = parser.parse_args()
+
+    check_regs(args.lhs_regs, args.lhs_bits)
+    check_regs(args.rhs_regs, args.rhs_bits)
+    check_regs(args.result_regs, args.lhs_bits + args.rhs_bits)
+
+    # Check that result registers don't overlap with operand registers.
+    for reg in args.result_regs:
+        if reg in args.lhs_regs or reg in args.rhs_regs:
+            raise ValueError(
+                f'Result register {reg} also appears in one of the operands.')
+
+    # Check that neither operand is 0 bits.
+    if args.lhs_bits == 0 or args.rhs_bits == 0:
+        raise ValueError('Both operands must have at least 1 bit.')
```

Comment:
Does this actually work with `lhs_bits` and `rhs_bits` equal to one?! Wow!

Reply:
Yep, it rounds the number of limbs up to the nearest 64-bit mark!
```shell
jadep@jadep-l:~/opentitan/hw/ip/otbn/util$ ./make_mulqacc.py --lhs-bits 1 --rhs-bits 1 --lhs-regs w2 --rhs-regs w3 --result-regs w10
bn.mulqacc.wo.z  w10, w2.0, w3.0
```